analyzers/src/SonarAnalyzer.CSharp/Rules/MultilineBlocksWithoutBrace.cs

```diff
@@ -28,143 +28,117 @@
-            //{
-            //    if (currentIfStatement.Else == null)
-            //    {
-            //        return currentIfStatement.Statement;
-            //    }
-
-            //    statement = currentIfStatement.Else.Statement;
-            //    currentIfStatement = statement as IfStatementSyntax;
-            //}
-
-            //return statement;
+            statement = ifStatement.Else.Statement;
+            ifStatement = statement as IfStatementSyntax;
         }
+        return statement;
+    }
 
-        private static void CheckStatement(SyntaxNodeAnalysisContext context, StatementSyntax statement,
-            string executed, string execute)
+    private static void CheckStatement(SyntaxNodeAnalysisContext context, StatementSyntax first, string executed, string execute)
+    {
+        if (SecondStatement(context.Node, first) is { } second && IsNotEmpty(first) && IsNotEmpty(second))
         {
-            if (statement.IsKind(SyntaxKind.Block))
-            {
-                return;
-            }
+            var firstPosition = ReferencePosition(first);
+            var secondPosition = StartPosition(second);
 
-            var nextStatement = context.Node.GetLastToken().GetNextToken().Parent;
-            // This algorithm to get the next statement can sometimes return a parent statement (for example a BlockSyntax)
-            // so we need to filter this case by returning if the nextStatement happens to be one ancestor of statement.
-            if (nextStatement == null ||
-                statement.Ancestors().Contains(nextStatement))
+            if (secondPosition.Character > firstPosition.Character - 4)
             {
-                return;
+                var lineSpan = context.Node.SyntaxTree.GetText().Lines[secondPosition.Line].Span;
+                var location = Location.Create(context.Node.SyntaxTree, TextSpan.FromBounds(second.SpanStart, lineSpan.End));
+                var blockSize = secondPosition.Line - StartPosition(first).Line + 1;
+                var additional = new[] { first.GetLocation() };
+                context.ReportIssue(Rule.CreateDiagnostic(context.Compilation, location, additional, executed, execute, blockSize));
             }
+        }
+    }
 
-            var statementPosition = statement.GetLocation().GetLineSpan().StartLinePosition;
-            var nextStatementPosition = nextStatement.GetLocation().GetLineSpan().StartLinePosition;
+    private static bool IsNotEmpty(SyntaxNode node) =>
+        node is not EmptyStatementSyntax;
 
-            if (statementPosition.Character == nextStatementPosition.Character)
-            {
-                var lineSpan = context.Node.SyntaxTree.GetText().Lines[nextStatementPosition.Line].Span;
-                var location = Location.Create(context.Node.SyntaxTree, TextSpan.FromBounds(nextStatement.SpanStart, lineSpan.End));
+    private static LinePosition ReferencePosition(SyntaxNode node)
```

Comment:
Having the refactoring below, can we come up with more descriptive name that will understand what's going on?

The current name is nicely short. But it takes a minute to understand what it really does

Reply:
@pavel-mikula-sonarsource This was my best attempt so far,as `PositionOfSelfOrAncestorsWhileOnSameLine()` felt a bit verbose. Maybe `ReferenceStartPosition()` is more descriptive. I'm not sure. (Naming things and cache invalidation...)

Reply:
What about `(Reference)IndentationStartPosition`? As that is what we want to find: a start position that returns some indentation to compare with. The only reason I return the `LinePosition`, and not the `Character` is that I also re-use the line in the report.